utils/events.py

```diff
@@ -135,3 +153,305 @@ def ensure_event_happens(self, target_type, target_id, event_type):
         if len(events) == 0:
             raise AssertionError(
                 'Event {}/{}/{} did not happen.'.format(event_type, target_type, target_id))
+
+
+class EventBuilder(object):
+    """
+    this class just simplifies "expected" event creation.
+
+    Usage:
+        builder = EventBuilder()
+        evt = builder.new_event(target_type='VmOrTemplate',
+                                target_name='my_lovely_vm',
+                                event_type='vm_create')
+    """
+    def __init__(self):
+        self._tool = EventTool(get_or_create_current_appliance())
+
+    def new_event(self, *attrs, **kwattrs):
+        event = Event(event_tool=self._tool)
+        for name, value in kwattrs.items():
+            event.add_attrs(EventAttr(**{name: value}))
+
+        for attr in attrs:
+            event.add_attrs(EventAttr(**attr))
+        return event
+
+
+class EventAttr(object):
+    """
+    contains one event attribute and the method for comparing it.
+    """
+    def __init__(self, attr_type=None, cmp_func=None, **attrs):
+        if len(attrs) > 1:
+            raise ValueError('event attribute can have only one key=value pair')
+
+        self.name, self.value = attrs.items()[-1]
+        self.type = attr_type or type(self.value)
+        self.cmp_func = cmp_func
+
+    def match(self, attr):
+        """
+        compares current attribute with passed attribute
+        """
+        if not isinstance(attr, EventAttr) or self.name != attr.name:
+            raise ValueError('Incorrect attribute is passed')
+
+        if attr.value is None or self.value is None:
+            return attr.value is None and self.value is None
+        elif self.cmp_func is not None:
+            return self.cmp_func(self.value, attr.value)
+        else:
+            return self.value == attr.value
+
+    def __repr__(self):
+        return "{name}({type})={val}, cmp_func {cmp}".format(name=self.name, type=self.type,
+                                                             val=self.value, cmp=self.cmp_func)
+
+
+# fixme: would it be better to create event prototype and just clone it ?
+class Event(object):
+    """
+    represents either db event received by CFME and stored in event_streams or an expected event
+    """
+    def __init__(self, *args, **kwargs):
+
+        # preparing connection to appliance db and filling metadata
+        if 'event_tool' in kwargs:
+            self._tool = kwargs.pop('event_tool')
+        else:
+            raise ValueError("event tool is required to create event")
+
+        # filling obtaining default attributes and their types
+        self._default_attrs = {}  # EventAttr obj
+        self._populate_defaults()
+
+        # container for event attributes
+        self.event_attrs = {}  # EventAttr obj
+
+        if len(args) > 0:
```

Comment:
nice catch. I changed this code several times and this is one of old statements I'd missed.